packages/request-state/src/runtime/middleware.ts

```diff
@@ -26,13 +26,13 @@ function injectState(getState: () => string | false) {
 	return new TransformStream({
 		transform(chunk, controller) {
 			if (!injected) {
-				const bodyCloseIndex = chunk.indexOf('</body>');
-				if (bodyCloseIndex > -1) {
+				const headCloseIndex = chunk.indexOf('</head>');
+				if (headCloseIndex > -1) {
```

Comment:
> This patch fixes middleware logic that inserts hydration data into end of head rather than end of body with dynamic scripts imports.

The state from the server hasn't been entirely defined when the end of head is sent due to response streaming. For example, think of this Astro page:
```astro
---
import { getState, setState } from '@it-astro:state';

const items = [1,2,3];
---
<ul>
  {items.map(async item => {
    const name = await getItemName(item);
	setState(`item-${item}`, name);
    return <li>{item}: {name}</li>;
  })}
</ul>
```

By the time `</head>` is sent the state hasn't been finalized, so the value reaching the client would be incomplete. To add the state before `</head>`, we'd need to collect the entire response body to ensure all the state has been processed and then modify it. This would effectively disable response streaming, which would be a huge performance hit for Astro apps using it.

We should either:
1. Document that the state is only available after `DOMContentLoaded` or `load` events (both are triggered only after the state loads) in the same way we document the relations with the View Transitions events;
2. Keep this change to add the state before `</head>` and further change to prevent streaming. This implies:
  - A breaking release since it is a breaking change,
  - Documenting that it disables streaming as we do for [Portal Gun](https://inox-tools.fryuni.dev/portal-gun#caveats)
3. Put this change behind a configuration so users can opt for streaming or early available state.

Since the package is not 1.0 yet, I'm okay with a breaking change to disable streaming. Adding back support for it would not be a breaking change, so it is fine to do it later with a more robust approach.

I'll leave it to you to decide which of the 3 paths above you'd want to take with your PR. All of them are acceptable to me.

Reply:
I think the easiest solution to this problem with streaming would be giving end user some sort of state hydration component that whoud replaced in middleware with actual data, simular to `NextScript` from `next/document`.

```astro
---
import { StateHydrationPoint } from "@it-astro:state"
---
<html>
  <head>
     <StateHydrationPoint/>
  </head>
  <body class="dark">
    hello
  </body>
</html>
```

Reply:
@MWGuy 

I use [patch-package](https://github.com/ds300/patch-package) to implement this solution in my project. I always set my shared nanostores at page or middleware level (before rendering the "it-astro-state" script).

`src/layouts/layout.astro`:
```astro
<head>
    <script
      is:inline
      id="it-astro-state"
      set:html={Astro.locals.astroState.getState()}
    />
</head>
```

`patches/@inox-tools+request-state+0.1.4.patch`:
```patch
diff --git a/node_modules/@inox-tools/request-state/dist/runtime/middleware.js b/node_modules/@inox-tools/request-state/dist/runtime/middleware.js
index 6b9f008..6306775 100644
--- a/node_modules/@inox-tools/request-state/dist/runtime/middleware.js
+++ b/node_modules/@inox-tools/request-state/dist/runtime/middleware.js
@@ -1,8 +1,12 @@
-import { defineMiddleware } from 'astro/middleware';
-import { collectState } from './serverState.js';
-import { parse } from 'content-type';
+import { defineMiddleware } from "astro/middleware";
+import { collectState, createState } from "./serverState.js";
 
-const u=defineMiddleware(async(i,n)=>{const{getState:e,result:t}=await collectState(n),r=t.headers.get("Content-Type");if(r===null)return t;const{type:o}=parse(r);if(o!=="text/html"&&!o.startsWith("text/html+"))return t;const s=t.body?.pipeThrough(new TextDecoderStream).pipeThrough(d(e)).pipeThrough(new TextEncoderStream);return new Response(s,t)});function d(i){let n=!1;return new TransformStream({transform(e,t){if(!n){const r=e.indexOf("</body>");if(r>-1){const o=i();if(o){const s=`<script id="it-astro-state" type="application/json+devalue">${o}</script>`;e=e.slice(0,r)+s+e.slice(r);}n=!0;}}t.enqueue(e);}})}
+const u = defineMiddleware(async (i, n) => {
+  const createdState = createState();
+  i.locals.astroState = createdState;
+  const result = await collectState(n, createdState);
+  return result;
+});
 
 export { u as onRequest };
 //# sourceMappingURL=middleware.js.map
diff --git a/node_modules/@inox-tools/request-state/dist/runtime/serverState.js b/node_modules/@inox-tools/request-state/dist/runtime/serverState.js
index 95e486d..1ab4249 100644
--- a/node_modules/@inox-tools/request-state/dist/runtime/serverState.js
+++ b/node_modules/@inox-tools/request-state/dist/runtime/serverState.js
@@ -1,8 +1,42 @@
-import { AsyncLocalStorage } from 'node:async_hooks';
-import { stringify } from 'devalue';
+import { AsyncLocalStorage } from "node:async_hooks";
+import { stringify } from "devalue";
 
-const o=new AsyncLocalStorage,l=(t,e)=>{const n=o.getStore();if(n!==void 0)return n.has(t)||e!==void 0&&n.set(t,e),n.get(t)},p=(t,e)=>{const n=o.getStore();e===void 0?n?.delete(t):n?.set(t,e);},i=new Map(Object.entries(Symbol).filter(([t,e])=>typeof e=="symbol"&&typeof t=="string").map(([t,e])=>[e,t])),c={URL:t=>t instanceof URL&&t.href,Date:t=>t instanceof Date&&t.valueOf(),GlobalSymbol:t=>typeof t=="symbol"&&t.description!==void 0&&t===Symbol.for(t.description)&&t.description,WellKnownSymbol:t=>typeof t=="symbol"&&i.get(t)},d=async t=>{const e=new Map;return {result:await o.run(e,t),getState:()=>e.size>0&&stringify(e,c)}};
+const o = new AsyncLocalStorage(),
+  l = (t, e) => {
+    const n = o.getStore();
+    if (n !== void 0)
+      return n.has(t) || (e !== void 0 && n.set(t, e)), n.get(t);
+  },
+  p = (t, e) => {
+    const n = o.getStore();
+    e === void 0 ? n?.delete(t) : n?.set(t, e);
+  },
+  i = new Map(
+    Object.entries(Symbol)
+      .filter(([t, e]) => typeof e == "symbol" && typeof t == "string")
+      .map(([t, e]) => [e, t])
+  ),
+  c = {
+    URL: (t) => t instanceof URL && t.href,
+    Date: (t) => t instanceof Date && t.valueOf(),
+    GlobalSymbol: (t) =>
+      typeof t == "symbol" &&
+      t.description !== void 0 &&
+      t === Symbol.for(t.description) &&
+      t.description,
+    WellKnownSymbol: (t) => typeof t == "symbol" && i.get(t),
+  },
+  createState = () => {
+    const state = new Map();
+    return {
+      state,
+      getState: () => stringify(state, c),
+    };
+  },
+  d = (t, createdState) => {
+    return o.run(createdState.state, t);
+  };
 
-export { d as collectState, l as getState, p as setState };
+export { d as collectState, l as getState, p as setState, createState };
 //# sourceMappingURL=serverState.js.map
 //# sourceMappingURL=serverState.js.map
\ No newline at end of file
```

Reply:
A state component would not solve the streaming problem, which make even state set on the front matter of components unreliable.

Only state set from the front matter of the _page_ would be guaranteed to have been set.

Another problem with a state component is that components wouldn't work by themselves. The page/layout that uses the component would need to know that some of it's components are implemented using request state and be modified accordingly.

A library providing a component, for example, would require the consumer to make changes to the layout of the page just to use a component somewhere else.

Using it on Starlight would require overriding the head component to add it there instead of just being available for your components. 

As I mentioned is fine too keep this change. Between a state component and just disabling streaming I'd prefer to disable streaming.